src/senaite/sync/complementstep.py

```diff
@@ -37,28 +38,74 @@ def _fetch_data(self):
         logger.info("*** COMPLEMENT STEP - FETCHING DATA: {} ***".format(
             self.domain_name))
 
-        self.uids = []
+        self.records = []
+        self.waiting_records = []
         self.sh = SoupHandler(self.domain_name)
-
-        # TODO: Find another way to do it without waking up objects.
+        # Dummy query to get overall number of items in the specified catalog
         query = {
             "url_or_endpoint": "search",
             "catalog": 'uid_catalog',
-            "b_start": 0,
-            "complete": "yes",
-            "limit": 500
+            "limit": 1
         }
         if self.content_types:
             query["portal_type"] = self.content_types
-        items = self._yield_items(**query)
-        for item in items:
-            # skip object or extract the required data for the import
-            if not item or not item.get("portal_type", True):
-                continue
-            data_dict = utils.get_soup_format(item)
-            rec_id = self.sh.insert(data_dict)
-            self.uids.insert(0, data_dict[REMOTE_UID])
+        cd = self.get_json(**query)
+        # Knowing the catalog length compute the number of pages we will need
+        # with the desired window size and overlap
+        window = 500
+        overlap = 5
+        effective_window = window-overlap
+        number_of_pages = (cd["count"]/effective_window) + 1
+        # Retrieve data from catalog in batches with size equal to window,
+        # format it and insert it into the import soup
+        for current_page in xrange(number_of_pages):
+            start_from = (current_page * window) - overlap
+            query["complete"] = True
+            query["limit"] = window
+            query["b_start"] = start_from
+            items = self.get_items(**query)
+            # Try one more time if items are not retrieved
+            if not items:
```

Comment:
I suggest to add two variables at the top of the class to make this more flexible. Sometimes, depending on the infrastructure, one might find useful to be able to set the maximum number of retries and the sleep time in seconds between 2 retries as well. I suggest to set the variables with the following values by default:
```python
MAX_RETRIES=5
SLEEP_MS_RETRIES=5
```